lib/kazan/watcher.ex

```diff
@@ -0,0 +1,164 @@
+    send_to = Keyword.fetch!(opts, :send_to)
+    rv = Keyword.fetch!(opts, :resource_version)
+    GenServer.start_link(__MODULE__, [request, rv, send_to])
+  end
+
+  @impl GenServer
+  def init([request, rv, send_to]) do
+    Logger.info "Watcher init pid: #{inspect self()} rv: #{rv} request: #{inspect request}"
+    {:ok, id} = Kazan.Client.run(request, stream_to: self())
+    {:ok, %State{id: id, request: request, rv: rv, send_to: send_to}}
+  end
+
+  @impl GenServer
+  def handle_info(%HTTPoison.AsyncStatus{code: 200}, state) do
+    {:noreply, state}
+  end
+
+  @impl GenServer
+  def handle_info(%HTTPoison.AsyncHeaders{}, state) do
+    {:noreply, state}
+  end
+
+  @impl GenServer
+  def handle_info(%HTTPoison.AsyncChunk{chunk: chunk, id: request_id},
+                  %State{id: id, buffer: buffer, rv: rv, send_to: send_to} = state) do
+    if request_id != id do
+      Logger.warn "Received chunk for wrong request_id: #{inspect request_id} expect: #{inspect id}"
+      {:noreply, state}
+    else
+
+      {lines, buffer} =
+        buffer
+        |> LineBuffer.add_chunk(chunk)
+        |> LineBuffer.get_lines()
+
+        new_rv = process_lines(lines, rv, send_to)
+
+      {:noreply, %State{state | buffer: buffer, rv: new_rv}}
+    end
+  end
+
+  @impl GenServer
+  def handle_info(%HTTPoison.Error{reason: {:closed, :timeout}}, state) do
+    Logger.info "Received Timeout with id: #{inspect self()}"
+    {:noreply, restart_connection(state)}
+  end
+
+  @impl GenServer
+  def handle_info(%HTTPoison.AsyncEnd{id: request_id},
+                  %State{id: id} = state) do
+    Logger.info "Received AsyncEnd for id: #{inspect request_id}"
+    if request_id != id do
+      Logger.warn "Received AsyncEnd message for wrong request_id: #{inspect request_id} expect: #{inspect id}"
+      {:noreply, state}
+    else
+      {:noreply, restart_connection(state)}
+    end
+  end
+
+
+  # INTERNAL
+
+  defp restart_connection(%State{request: request, rv: rv} = state) do
+    query_params = Map.put(request.query_params, "resourceVersion", rv)
+    request = %Kazan.Request{request | query_params: query_params}
+    {:ok, id} = Kazan.Client.run(request, stream_to: self())
+    Logger.info "Restarted #{inspect self()}"
+    %State{state | id: id, buffer: LineBuffer.new()}
+  end
+
+  defp process_lines(lines, rv, send_to) do
+    Enum.reduce(lines, rv, fn(line, current_rv) ->
+      {:ok, %{"type" => type, "object" => raw_object}} = Poison.decode(line)
+      {:ok, model} = Kazan.Models.decode(raw_object)
+      ve = %Kazan.Models.ApiMachinery.Apis.Meta.V1.WatchEvent{type: type, object: model}
+
+      send(send_to, ve)
+
+      new_rv(type, raw_object, current_rv)
+    end)
+  end
+
+  defp new_rv("DELETED", _object, current_rv), do: current_rv
+  defp new_rv(type, object, current_rv) do
+    case object do
+      %{"metadata" => %{"resourceVersion" => new_rv}} ->
+        if String.to_integer(new_rv) > String.to_integer(current_rv) do
```

Comment:
The [kube docs](https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#concurrency-control-and-consistency) say

> It's important to note that the application should not rely on the implementation details of the versioning system maintained by Kubernetes. We may change the implementation of resourceVersion in the future, such as to change it to a timestamp or per-object counter.

and 

> This value MUST be treated as opaque by clients

So I don't think it's safe to treat it like an integer here.

Reply:
Not sure what the best thing to do here is.  I read this in the docs also, however I've seen that sometimes events do always arrive in the correct order, therefore if we just blindly update the resource version, when we have to restart the request, we can get old events again.

One approach could be to store a list of RVs of events already received, and then not resend these events if they are received again?

Reply:
However, both their alternative suggestions as to future implementations suggest that there is still ordering.

Reply:
Hmm.  I'm wondering if this out of order messages thing is expected behaviour, or a bug in k8s.  If it's a bug then I'd definitely prefer we don't work around it ourselves.  If it's expected behaviour, then I wonder if there's a recommended way to deal with it...

Reply:
I'll investigate further as to when it happens and let you know.

Reply:
After some investigation, I have reproduced the problem with events being received out of order.  It is in fact confirmed as a bug in K8S and caused by the `DELETE` events sending the wrong RV:

https://github.com/kubernetes/kubernetes/issues/58545

The simplest workaround, I believe, is to not update the RV held in the state with the RV from a `DELETED` event.  This will ensure that when the connection restarts, we don't use an incorrect RV.  Then the code does not to do any comparison of the RV to check if it is more recent or not.

However, any `DELETED` events received after the last `ADDED` or `MODIFIED` events before the connection is restarted, will be resent by K8S to the Watcher after restart as it does not have the correct RV for the new request.  The Watcher will then send these duplicates on to the client.

To solve the Watcher could either:
- Store all the `DELETED` events received after a `ADDED` or `MODIFIED` so it can detect duplicates and not send them on to the client.
- Just document that clients may receive duplicate `DELETED` events and the client code should be idempotent for handling them.

What do you think?

Reply:
Nice - thanks for looking into that.  I tried to find an issue in the k8s repo myself, but didn't have much luck finding anything.

I don't really think we should work around this in Kazan.  It was definitely an upstream issue, and the issue is already fixed in k8s.  I don't think it's the job of a client library to try and hide bugs in upstream servers.  It seems like this is also the approach the go client library has taken.

Reply:
OK - sure.  I will not workaround it.

Hopefully a 1.8 version of K8S with this fix will be out within a couple of weeks allowing me to test this PR properly.  Unfortunately without the fix or workaround, this Watch functionality a bit unusable, at least for my use case.

Reply:
I was in a bit of a rush when writing that last comment, so there's some explanation I left out:

Part of why I'm uneasy about adding this workaround to kazan is that as library authors we're not well placed to figure out whether our users actually need the workaround.  A user could easily be using a fixed version of k8s, in which case the workaround of ignoring rvs for `DELETED` events would be the wrong behavior.  Then there's the flipside of that: we can't really know when users have actually upgraded to a fixed version of k8s, so it's unclear when it'd be suitable to remove the workaround from the kazan codebase.

The correct fix to make may also somewhat depend on a users use case - maybe the code that makes use of the watch events is idempotent, so duplicate events are fine.  Maybe they know events won't happen that often, so tracking resource versions to remove duplicates is fine.  Or maybe they know they'll be getting a ton of updates, so tracking resource versions is not going to be workable.

Essentially, I think a user of kazan is a lot better placed to decide what the correct thing to do is - so it makes more sense for workarounds for this issue to be done in the applications that use kazan, rather than in kazan itself.

I don't know what your use case is, but would it not be possible to implement the workaround in your application?  Presumably you could just have a GenServer that starts up a Watcher, gets events from it, filters out duplicates and sends the rest on to some final destination...

Reply:
Yes, I agree you are right.  I think not adding workarounds is the best approach.  

I'm not sure if the workaround would be that simple to implement in the application, as the list of events that would need storing would be unbounded.

However, as I said earlier, I will wait a couple of weeks until the hopefully the bug fix is released, then I can finish the PR, verify it works correctly and also add an integration test.  I can also document the issue so users are aware of this if they are using an earlier version of K8S.